### PR TITLE
Flipping origin-spanning locations with strand=None

### LINKED ISSUE #4611: Flipping origin-spanning locations with strand=None in circular molecules
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```python
import sys; print(sys.version)
import platform; print(platform.python_implementation()); print(platform.platform())
import Bio; print(Bio.__version__)
```

```
3.11.6 (main, Nov  2 2023, 04:51:19) [Clang 14.0.0 (clang-1400.0.29.202)]
CPython
macOS-12.6.7-x86_64-i386-64bit
1.81
```
In Biopython and GenBank files, locations that span the origin are represented by a compound location. For example, for a sequence of length 6 a feature that spans two residues before the origin and the first residue after the origin is represented as `join{[4:6](+), [0:1](+)}`. When we flip that feature, it becomes `join{[0:2](-), [5:6](-)}`. The order of the loc.parts and the strand is inverted, so the meaning of the flipping is clear.

However, when a feature without strand spans the origin in the same way (`join{[4:6], [0:1]}`) and gets flipped, the parts get inverted (`join{[0:2], [5:6]}`). It might be confusing to interpret that this is a feature that spans the origin.

The same applies to multi-part features, see example below:

```python
from Bio.SeqFeature import SimpleLocation, SeqFeature
from Bio.SeqRecord import SeqRecord
from Bio.Seq import Seq
from Bio.SeqIO import write
loc = SimpleLocation(4, 6, 1) + SimpleLocation(0, 1, 1)
print(loc) # join{[4:6](+), [0:1](+)}
print(loc._flip(6)) # join{[0:2](-), [5:6](-)}

loc = SimpleLocation(4, 6, None) + SimpleLocation(0, 1, None)
print(loc) # join{[4:6], [0:1]}
print(loc._flip(6)) # join{[0:2], [5:6]}


# Example of a multi-part location that spans the origin of a sequence with length 20
loc = SimpleLocation(15, 16, 1) + SimpleLocation(18, 20, 1) + SimpleLocation(0, 1, 1) + SimpleLocation(5, 10, 1)
print(loc) # join{[15:16](+), [18:20](+), [0:1](+), [5:10](+)}
print(loc._flip(20)) # join{[4:5](-), [0:2](-), [19:20](-), [10:15](-)}
seqr = SeqRecord(Seq('A'*20), features=[SeqFeature(loc, type='misc_feature')], annotations={'molecule_type': 'DNA'})
write([seqr], 'example_strand1_fwd.gb', 'genbank')
seqr_rvs = seqr.reverse_complement()
seqr_rvs.annotations['molecule_type'] = 'DNA'
write([seqr_rvs], 'example_strand1_rvs.gb', 'genbank')

loc = SimpleLocation(15, 16, None) + SimpleLocation(18, 20, None) + SimpleLocation(0, 1, None) + SimpleLocation(5, 10, None)
print(loc) # join{[15:16], [18:20], [0:1], [5:10]}
print(loc._flip(20)) # join{[4:5], [0:2], [19:20], [10:15]}
seqr = SeqRecord(Seq('A'*20), features=[SeqFeature(loc, type='misc_feature')], annotations={'molecule_type': 'DNA'})
seqr_rvs = seqr.reverse_complement()
seqr_rvs.annotations['molecule_type'] = 'DNA'
write([seqr], 'example_strandNone_fwd.gb', 'genbank')
write([seqr_rvs], 'example_strandNone_rvs.gb', 'genbank')

```
In the example, I save the sequences with multi-part features to gb files. As you can see, snapgene interprets properly all cases, except the inverted feature without strand. You can access [gb files here](https://github.com/biopython/biopython/files/14111421/files.zip).

Screenshot 2024-01-31 at 12 06 38

If you agree that this is something that should be changed, I can give it a go and propose a solution.

### PR DESCRIPTION
**Pull request by manulera**
- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

Closes #4611

### CODE REVIEW

File: Bio/SeqFeature.py [hunk 1]
```diff
@@ -1641,7 +1641,7 @@ def _shift(self, offset):
     def _flip(self, length):
         """Return a copy of the locations after the parent is reversed (PRIVATE).
 
-        Note that the order of the parts is NOT reversed too. Consider a CDS
+        Note that the order of the parts is NOT reversed if strand!= None. Consider a CDS
```
Comment (peterjc) [inline Bio/SeqFeature.py:1644]: How about "Note that the order of the parts is NOT reversed if any of the parts are stranded", and reword the line wrapping?

File: Bio/SeqFeature.py [hunk 2]
```diff
@@ -1709,7 +1709,16 @@ def _flip(self, length):
         versions of Biopython which would have given join{[5:29](-), [37:52](-)}
         and the translation would have wrongly been "EXAMPLE*SILLY" instead.
 
+        When strand is None, the parts are reversed. In principle this does not
```
Comment (peterjc) [inline Bio/SeqFeature.py:1712]: How about "When all the parts have strand None, the order of the parts is reversed..."?

File: Bio/SeqFeature.py [hunk 3]
```diff
@@ -1709,7 +1709,16 @@ def _flip(self, length):
...
+        change the meaning of the location, but improves the representation of
+        locations spanning the origin in circular molecules. For more details,
+        see https://github.com/biopython/biopython/issues/4611
```
Comment (peterjc) [inline Bio/SeqFeature.py:1715]: I would leave out the issue reference, I don't think this adds anything to the documentation.

Perhaps insert "misc-features" though, as in "... improves the representation of misc-features spanning the origin in circular molecules"?

File: Bio/SeqFeature.py [hunk 4]
```diff
@@ -1709,7 +1709,16 @@ def _flip(self, length):
...
+
         """
+        if all(loc.strand is None for loc in self.parts):
+            return CompoundLocation(
+                [loc._flip(length) for loc in self.parts[::-1]], self.operator
+            )
```
Comment (peterjc) [inline Bio/SeqFeature.py:1721]: I would find an "else" clearer here

File: NEWS.rst [hunk 5]
```diff
@@ -80,6 +80,10 @@ The UniProt package now includes a method to search UniProt programmatically.
 The tutorial has been updated with examples of how to use the search method.
 See the chapter on Swiss-Prot and ExPASy in the tutorial for more information.
 
+Now when "flipping" compound locations with ``_flip()`` in ``Bio.SeqFeature``,
+if the strand is ``None``, the order of the sublocations is reversed, to improve
+the representation of origin-spanning locations. See issue #4611.
+
```
Comment (peterjc) [inline NEWS.rst:86]: Lead with taking the reverse complement of a SeqRecord, the fact the change is in a private method ``._flip()`` probably doesn't need mentioning.

**Review by peterjc — CHANGES_REQUESTED** [review]
The code and test look good. I'd like to improve the documentation side of things though.

File: Bio/SeqFeature.py (hunk 3 above)
Comment (manulera) [reply Bio/SeqFeature.py:1715]: I have done something in these lines, but I don't this is restricted to misc-features. Could be a centromere, for instance.

**Comment by manulera** [comment]
I have addressed those requested changes in the last commit

**Comment by peterjc** [comment]
That's great - thank you. Apologies I didn't get to this sooner.

**Comment by manulera** [comment]
No problem, thanks for merging. Have a nice weekend

**Pull request merged**